File: csharp/SampleGameBackend/Controllers/ValidationController.cs
using Microsoft.AspNetCore.Mvc;
using SampleGameBackend.Database;
using SampleGameBackend.Models;
using SampleGameBackend.Services;
using System.Text.Json;

namespace SampleGameBackend.Controllers
{
    [ApiController]
    [Route("api")]
    public class ValidationController : ControllerBase
    {
        private readonly ValidationService _validationService;
        private readonly DatabaseService _databaseService;
        private readonly ILogger<ValidationController> _logger;

        public ValidationController(ValidationService validationService, DatabaseService databaseService, ILogger<ValidationController> logger)
        {
            _validationService = validationService;
            _databaseService = databaseService;
            _logger = logger;
        }

        [HttpPost("validate")]
        public IActionResult ValidateUserAction([FromBody] ValidateRequest req)
        {
            // TODO: We need a defense logic to prevent duplicate UUIDs in requests.
            try
            {
                // Request validation
                if (req == null || string.IsNullOrEmpty(req.Uuid) || string.IsNullOrEmpty(req.UserSig) || 
                    string.IsNullOrEmpty(req.UserAddress) || string.IsNullOrEmpty(req.ProjectId) || 
                    string.IsNullOrEmpty(req.Digest) || req.Intent == null)
                {
                    return BadRequest(new ValidateResponse
                    {
                        Success = false,
                        ErrorCode = "INVALID_REQUEST"
                    });
                }

                // Intent validation
                if (!_validationService.ValidateIntent(req.Intent))
                {
                    return BadRequest(new ValidateResponse
                    {
                        Success = false,
                        ErrorCode = "INVALID_INTENT"
                    });
                }

                // Get session ID
                var sessionId = HttpContext.Items["SessionId"] as string;
                if (string.IsNullOrEmpty(sessionId))
                {
                    return BadRequest(new ValidateResponse
                    {
                        Success = false,
                        ErrorCode = "INVALID_SESSION"
                    });
                }

                // Store UUID and SessionID mapping
                try
                {
                    _databaseService.StoreUuidMapping(req.Uuid, sessionId);
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "Failed to store UUID mapping");
                    return StatusCode(500, new ValidateResponse
                    {
                        Success = false,
                        ErrorCode = "UUID_MAPPING_FAILED"
                    });
                }

                var requestJson = JsonSerializer.Serialize(req);
                _logger.LogInformation("ValidateUserActionHandler: sessionID={SessionId}, uuid={Uuid}, req={Request}", 
                    sessionId, req.Uuid, requestJson);

                // For mint method, validate and deduct assets
                if (req.Intent.Method == "mint" || req.Intent.Method == "transfer")
                {
                    try
                    {
                        _validationService.ValidateAndProcessMint(sessionId, req.Intent.From);
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Insufficient balance for sessionId: {SessionId}", sessionId);
                        return BadRequest(new ValidateResponse
                        {
                            Success = false,
                            ErrorCode = "INSUFFICIENT_BALANCE"
                        });
                    }
                }

                // Generate validator signature
                string validatorSig;
                try
                {
                    validatorSig = _validationService.GenerateValidatorSignature(req.UserSig, req.Digest);
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "GenerateValidatorSignature failed");
                    return StatusCode(500, new ValidateResponse
                    {
                        Success = false,
                        ErrorCode = "SIGNATURE_GENERATION"
                    });
                }

                _logger.LogInformation("validateUserActionHandler: validatorSig={ValidatorSig}, userSig={UserSig}, digest={Digest}", 
                    validatorSig, req.UserSig, req.Digest);

                // Success response
                var response = new ValidateResponse
                {
                    Success = true,
                    ErrorCode = null,
                    Data = new ValidateResponseData
                    {
                        UserSig = req.UserSig,
                        ValidatorSig = validatorSig
                    }
                };

                return Ok(response);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "ValidateUserAction error");
                return StatusCode(500, new ValidateResponse
                {
                    Success = false,
                    ErrorCode = "INTERNAL_ERROR"
                });
            }
        }
    }
} 

File: typescript/src/controllers/validationController.ts
import { Response } from 'express';
import { AuthenticatedRequest } from '../middleware/authMiddleware';
import { ValidateRequest, ValidateResponse } from '../types';
import { ValidationService } from '../services/validationService';
import { MemoryDatabase } from '../database/memoryDb';

export class ValidationController {
  private validationService: ValidationService;
  private db: MemoryDatabase;

  constructor(validationService: ValidationService, db: MemoryDatabase) {
    this.validationService = validationService;
    this.db = db;
  }

  /**
   * Validate user action
   * @param req - Request object
   * @param res - Response object
   */
  async validateUserAction(req: AuthenticatedRequest, res: Response): Promise<void> {
    try {
      const sessionId = req.sessionId;
      const requestBody = req.body as ValidateRequest;

      // Request validation
      if (!requestBody || !requestBody.uuid || !requestBody.user_sig || 
          !requestBody.user_address || !requestBody.project_id || 
          !requestBody.digest || !requestBody.intent) {
        res.status(400).json({
          success: false,
          errorCode: 'INVALID_REQUEST',
          message: 'Missing required fields'
        });
        return;
      }

      // Intent validation
      if (!this.validationService.validateIntent(requestBody.intent)) {
        res.status(400).json({
          success: false,
          errorCode: 'INVALID_INTENT',
          message: 'Invalid intent structure'
        });
        return;
      }

      // Get session ID
      if (!sessionId) {
        res.status(400).json({
          success: false,
          errorCode: 'INVALID_SESSION_ID',
          message: 'Session ID is required'
        });
        return;
      }

      // TODO: We need a defense logic to prevent duplicate UUIDs in requests.
      // Store UUID and SessionID mapping
      try {
        this.db.storeUuidMapping(requestBody.uuid, sessionId);
      } catch (error) {
        console.error('Failed to store UUID mapping:', error);
        res.status(500).json({
          success: false,
          errorCode: 'UUID_MAPPING_FAILED',
          message: 'Failed to store UUID mapping'
        });
        return;
      }

      console.log('ValidateUserActionHandler:', {
        sessionID: sessionId,
        uuid: requestBody.uuid,
        req: JSON.stringify(requestBody)
      });

      // For mint method, validate and deduct assets
      if (requestBody.intent.method === 'mint' || requestBody.intent.method === 'transfer') {
        try {
          this.validationService.validateAndProcessMint(sessionId, requestBody.intent.from);
        } catch (error) {
          console.error('Insufficient balance for sessionId:', sessionId, error);
          res.status(400).json({
            success: false,
            errorCode: 'INSUFFICIENT_BALANCE',
            message: 'Insufficient balance for operation'
          });
          return;
        }
      }

      // Generate validator signature
      let validatorSig: string;
      try {
        validatorSig = await this.validationService.generateValidatorSignature(requestBody.user_sig, requestBody.digest);
      } catch (error) {
        console.error('GenerateValidatorSignature failed:', error);
        res.status(500).json({
          success: false,
          errorCode: 'SIGNATURE_GENERATION_FAILED',
          message: 'Failed to generate validator signature'
        });
        return;
      }

      console.log('validateUserActionHandler:', {
        validatorSig,
        userSig: requestBody.user_sig,
        digest: requestBody.digest
      });

      // Success response
      const response: ValidateResponse = {
        success: true,
        errorCode: null,
        data: {
          userSig: requestBody.user_sig,
          validatorSig
        }
      };

      res.json(response);
    } catch (error) {
      console.error('ValidateUserAction error:', error);
      res.status(500).json({
        success: false,
        errorCode: 'INTERNAL_ERROR',
        message: 'Internal server error'
      });
    }
  }
} 